docs/data/data-grid/components/columns-panel/GridColumnsPanelTrigger.js

```diff
@@ -0,0 +1,28 @@
+import * as React from 'react';
+import { DataGrid, Grid } from '@mui/x-data-grid';
+import { useDemoData } from '@mui/x-data-grid-generator';
+import ViewColumnIcon from '@mui/icons-material/ViewColumn';
+
+function Toolbar() {
+  return (
+    <Grid.Toolbar.Root>
+      <Grid.ColumnsPanel.Trigger render={<Grid.Toolbar.Button />}>
```

Comment:
I expected the `render` prop to accept a function since it's an established pattern. Then I checked [Base UI docs](https://base-ui.com/components/react-collapsible/#overriding-default-components) and realized that they support both signatures. I like the flexibility of this approach 👍🏻.

Reply:
Out of curiosity, for this example, why `<Trigger render={<Button />}>[content]</Trigger>` and not `<Trigger render={<Button>[content]</Button>} />`?

Isn't this format just another (more complex) way of writing what MUI usually writes as `<Box as={Button}>`?

Reply:
> Out of curiosity, for this example, why `<Trigger render={<Button />}>[content]</Trigger>` and not `<Trigger render={<Button>[content]</Button>} />`?

I think it just comes down to preference, both are valid. The only time content would _need_ to be specified inside the Button in the `render` prop, is when props or state are being used to conditionally render content.

Reply:
> Isn't this format just another (more complex) way of writing what MUI usually writes as <Box as={Button}>?

Yes, it solves the same problem of element replacement.

There are a few reasons why we're considering this approach:
- It's simpler to

Reply:
> Isn't this format just another (more complex) way of writing what MUI usually writes as <Box as={Button}>?

Yes, it solves the same problem of element replacement.

There are a few reasons we're considering this approach:
  - It's much simpler to type. In your example, the Box has to do some TypeScript gymnastics to understand which props the Button accepts.
  - There are some limitations to the Material UI override pattern, for example, prop collisions:

    ```tsx
    interface CustomButtonProps {
      children: React.ReactNode;
      color: 'color-that-does-not-exist-on-mui-button';
    }

    function CustomButton({ children, color }: CustomButtonProps) {
      return (
        <button type="button" color={color}>
          {children}
        </button>
      );
    }

    function Toolbar() {
      return (
        <>
          /* ❌ With component prop - type error
          Type '"color-that-does-not-exist-on-mui-button"' is not assignable to type 'OverridableStringUnion<"inherit" | "primary" | "secondary" | "success" | "error" | "info" | "warning", ButtonPropsColorOverrides> | undefined'. */
          <Button
            component={CustomButton}
            color="color-that-does-not-exist-on-mui-button"
          />

          /* ✅ With render prop - works correctly */
          <Button
            render={<CustomButton color="color-that-does-not-exist-on-mui-button" />} 
          />
        </>
      );
    }
    ```
    
    The render prop approach works because props are passed directly to the custom component.

- Base UI spent some time exploring the options for [element replacement](https://www.notion.so/mui-org/Component-Replacement-Showdown-128cbfe7b6608069b3c0d64321760d08) (not sure if this is the most up-to-date reference) and settled on this approach.
- X is also exploring the `render` function pattern, which is documented [here](https://www.notion.so/mui-org/Element-override-patterns-121cbfe7b66080efa637e41a1063343f). It feels like a good time to explore this, alongside the new composition pattern.

Reply:
Makes sense, thanks :)